active/001-tabular-msa.md

```diff
@@ -0,0 +1,292 @@
+acyclic graph, or DAG). Thus, while some operations are shared among these MSA
+structures, there are intrinsic differences that necessitate different APIs.
+
+In scikit-bio 0.4.0 the sequence classes were completely redesigned and the
+current MSA class (`Alignment`) makes invalid assumptions about the API of the
+sequences it stores (for example, see [#1036](https://github.com/biocore/scikit-bio/issues/1036)). This gives us an
+opportunity to rework the MSA API to better fit the new sequence classes.
+
+After these changes, we will have an obvious way to introduce new MSA
+structures (e.g., partial-order MSA), better interaction with the new sequence
+API, pandas-style vectorization, and flexible, arbitrary metadata that can
+represent file formats such as
+[Stockholm](http://sonnhammer.sbc.su.se/Stockholm.html).
+
+# Detailed design
+
+A `TabularMSA` object is a mutable, monomorphic, ordered collection of
+`Sequence` objects that are of equal length. As its name implies, `TabularMSA`
+represents a table-like structure where the first dimension is sequences and
+the second dimension is positions in the alignment. Thus, each row corresponds
+to a sequence and each column corresponds to a position. The first dimension is
+optionally indexed by an arbitrary key. Both dimensions are indexable via
+ordinals.
+
+One of the advantages of `TabularMSA` over the existing `Alignment` class is
+vectorization support and boolean indexing. Currently, to filter out alignment
+positions that have a Shannon entropy greater than 0.5:
+
+```python
+aln.subalignment(positions_to_keep=[
+    i for i, v in enumerate(aln.position_entropies()) if v <= 0.5])
+```
+
+With `TabularMSA`:
+
+```python
+aln[:, aln.entropies() <= 0.5]
+```
+
+Another example is filtering out completely-gapped positions. With the current
+API, this will (probably) work:
+
+```python
+aln.omit_gap_positions(1.0 - np.finfo(float).eps)
+```
+
+With `TabularMSA`:
+
+```python
+aln[:, aln.gap_frequencies(relative=True) < 1.0]
+```
+
+Thus with boolean vector support, `TabularMSA` is significantly more flexible
+in its filtering capabilities.
+
+## Constructors
+- `TabularMSA(iterable, key=None)`
+
+   Construct a `TabularMSA` from an iterable of `Sequence` objects. `key` is a
+   callable or a lookup key in `Sequence.metadata`. When `key=None`, will be
+   the ordinals 0 through N - 1.
+
+## Properties
+- `dtype -> Sequence type`
+
+   Type of sequence stored (e.g., `DNA`)
+- `shape -> (int, int)`
+
+   Tuple storing number of rows (sequences) and columns (positions)
+- `loc -> indexer`
+
+   Allows indexing via keys (like pandas)
+- `iloc -> indexer`
+
+   Allows indexing via indices (like pandas)
+
+- `metadata -> dict`
+
+   Arbitrary metadata that applies to the alignment as a whole.
+
+- `positional_metadata -> pd.DataFrame`
+
+   Arbitrary metadata that applies to the positions of the alignment as a
+   whole.
+
+## Methods
+- `__nonzero__() -> bool`
```

Comment:
So here's a thought on that:
Presently, the sequence axis is a monoid, having identity and associativity:
```
empty + aln = aln
aln + empty = aln
(aln1 + aln2) + aln3 = aln1 + (aln2 + aln3)
```
I think the key to this question is the identity object.
IF we have a `merge` operation that acts across positions instead of sequences, then we have our binary operator and to fulfill the monoid, we would expect an identity.

Presently the identity for the sequence axis is falsey, so then the identity for the position axis should also be falsey.

So then by this logic, and alignment with entirely empty sequences would also be falsey and our secondary monoid would look like this:
```
aln.merge(empty_seqs) == aln
empty_seqs.merge(aln) == aln
aln.merge(aln2).merge(aln3) == aln.merge(aln2.merge(aln3))
```